src/components/Navigator/QuickNavigationModal.vue

```diff
@@ -272,6 +343,78 @@ export default {
+      // data hasn't completed yet, indicate that the data is still loading in
+      // the actual UI — without the delay, the constant flashing of this
+      // message would be too distracting
+      this.loadingTimeout = setTimeout(() => {
+        if (this.previewState === PreviewState.loading) {
+          this.previewIsLoadingSlowly = true;
+        }
+      }, SLOW_LOADING_DELAY);
+
+      // exit early if the previwe data is already cached for this symbol as
+      // there is no additional work needed
+      if (!this.selectedSymbol || this.$cachedSymbolResults.has(this.selectedSymbol.uid)) {
+        clearTimeout(this.loadingTimeout);
+        this.previewIsLoadingSlowly = false;
+        return;
+      }
+
+      const fetchSymbolData = async (symbol) => {
+        if (!symbol || this.$cachedSymbolResults.has(symbol.uid)) {
+          return;
+        }
+
+        try {
+          // load render JSON for the selected /documentation/ page
+          const json = await fetchDataForPreview(symbol.path, {
+            signal: this.abortController.signal,
+          });
+          this.$cachedSymbolResults.set(symbol.uid, {
+            success: true,
+            json,
+          });
+        } catch (e) {
+          // errors triggered by the abort controller are safe to ignore since
+          // we are only aborting them for performance reasons and would like
+          // to later re-try them if possible
+          if (e.name !== ABORT_ERROR_NAME) {
+            this.$cachedSymbolResults.set(symbol.uid, {
+              success: false,
+            });
+          }
+        } finally {
+          // `LRUMap` is a custom object that is very similar to a builtin `Map`
+          // and since `Map` objects can't directly hook into the reactivity
+          // in Vue 2, we need to convert this object into a plain `Object` every
+          // time it is mutated to workaround this lack of reactivity
+          this.cachedSymbolResults = Object.fromEntries(this.$cachedSymbolResults);
+        }
+      };
+
+      // the abort controller is useful for clients with slower networks—if it
+      // is taking a long time for page data to load, we don't want to create
+      // a bottleneck with many queued pending requests. this controller is used
+      // to cancel any pending requests that aren't for this symbol or the one
+      // immediately following it which can happen on a slow network if the
+      // user quickly scrolls through the list of results
+      if (this.abortController) {
+        this.abortController.abort();
+      }
+      this.abortController = new AbortController();
+
+      await Promise.all([
+        fetchSymbolData(this.selectedSymbol).finally(() => {
+          // the timeout for the loading message only applies to the currently
+          // selected symbol
+          clearTimeout(this.loadingTimeout);
+          this.previewIsLoadingSlowly = false;
+        }),
+        fetchSymbolData(this.nextSymbol),
```

Comment:
Do we await this resulting promise anywhere? Just checking, because if we did, then waiting for the next item to finish might not be necessary always.

Reply:
Yeah, on line 408.

I agree that it would be better to optimize this so that it only waits for the first one and not necessarily the next one, although I was a little unclear on exactly how to get that code right—it's a little tricky, because we need to make sure to update the UI after waiting for the first one but also not mess up the next one once it's been selected as the current one. I can see if I can optimize this at all.

Reply:
Oh, you mean waiting for this whole promise somewhere. I don't believe so, since it's an async watcher.